jmespath-core/src/main/java/io/burt/jmespath/function/CompareByFunction.java

```diff
@@ -0,0 +1,56 @@
+package io.burt.jmespath.function;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.ArrayList;
+
+import io.burt.jmespath.Adapter;
+import io.burt.jmespath.node.JmesPathNode;
+
+public class CompareByFunction extends JmesPathFunction {
+  private final int compareModifier;
+
+  public CompareByFunction(int compareModifier) {
+    super(2, 2);
+    this.compareModifier = compareModifier;
+  }
+
+  @Override
+  protected <T> T internalCall(Adapter<T> adapter, List<ExpressionOrValue<T>> arguments) {
+    T array = arguments.get(0).value();
+    JmesPathNode expression = arguments.get(1).expression();
+    if (arguments.get(0).isExpression()) {
```

Comment:
I think it would be nicer to either make the `null`-checks explicit here, or completely internalize it `ExpressionOrValue`, by letting `value()` and `expression()` throw errors if it was the wrong kind. The current logic where you first perform the operation, and then afterwards check whether it was acceptable seems really weird to me.